hibike/unit_tests/hibike_message_tests.py

```diff
@@ -0,0 +1,373 @@
+    """
+    @staticmethod
+    def gen_random_device_id_and_params():
+        """
+        Generate a random device ID and set of parameters.
+        """
+        device_id = hibike_message.uid_to_device_id(random_uid())
+        params = random_params(device_id)
+        return (device_id, params)
+
+    def test_idempotence(self):
+        """
+        Check that encoding and decoding a random set of parameters
+        is idempotent.
+        """
+        def assert_params_same(device_id, param_list):
+            """
+            Check that encoding and decoding PARAM_LIST produces
+            the same result.
+            """
+            encoded = hibike_message.encode_params(device_id, param_list)
+            decoded = hibike_message.decode_params(device_id, encoded)
+            self.assertEqual(sorted(decoded), sorted(param_list),
+                             "encoded/decoded param list {} not identical".format(param_list))
+
+        run_with_random_data(assert_params_same,
+                             self.gen_random_device_id_and_params, times=1000)
+
+
+class ParsingTests(unittest.TestCase):
+    """
+    Tests for parsing Hibike messages.
+    """
+    MESSAGE_GENERATORS = {packet_type: None for packet_type in hibike_message.MESSAGE_TYPES}
+    VALUE_GENERATORS = {key: None for key in hibike_message.PARAM_TYPES}
+    @classmethod
+    def setUpClass(cls):
+        cls.MESSAGE_GENERATORS["SubscriptionRequest"] = cls.gen_random_sub_request
+        cls.MESSAGE_GENERATORS["DeviceRead"] = cls.gen_random_device_read
+        cls.MESSAGE_GENERATORS["HeartbeatResponse"] = cls.gen_random_heartbeat_response
+        cls.MESSAGE_GENERATORS["SubscriptionResponse"] = cls.gen_random_sub_response
+        cls.MESSAGE_GENERATORS["DeviceData"] = cls.gen_random_device_data
+        cls.MESSAGE_GENERATORS["DeviceWrite"] = cls.gen_random_device_write
+        cls.VALUE_GENERATORS["bool"] = lambda: random.randrange(2) == 0
+        cls.VALUE_GENERATORS["uint8_t"] = lambda: random.randrange(256)
+        cls.VALUE_GENERATORS["int8_t"] = lambda: random.randrange(-128, 128)
+        cls.VALUE_GENERATORS["uint16_t"] = lambda: random.randrange(2 ** 16)
+        cls.VALUE_GENERATORS["int16_t"] = lambda: random.randrange(-(2 ** 15), 2 ** 15)
+        cls.VALUE_GENERATORS["uint32_t"] = lambda: random.randrange(2 ** 32)
+        cls.VALUE_GENERATORS["int32_t"] = lambda: random.randrange(-(2 ** 31), 2 ** 31)
+        cls.VALUE_GENERATORS["uint64_t"] = lambda: random.randrange(2 ** 64)
+        cls.VALUE_GENERATORS["int64_t"] = lambda: random.randrange(-(2 ** 63), 2 ** 63)
+        cls.VALUE_GENERATORS["float"] = random.random
+
+    @staticmethod
+    def encode_packet(packet):
+        """
+        Turn a HibikeMessage into an encoded packet.
+        """
+        fake_port = FakeSerialPort()
+        hibike_message.send(fake_port, packet)
+        return fake_port.drain()
+
+    @staticmethod
+    def gen_random_sub_request():
```

Comment:
would it perhaps be useful to make these family of functions take arguments, with default values set to something? (maybe `None`, though if you did that, to get the same cleanless you'd have to create another function to wrap the `if None, else random()` check)

Reply:
Hm, not quite sure why this would be better. More flexibility?